File: src/test/java/io/femo/http/HttpTest.java
package io.femo.http;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import io.femo.http.drivers.DefaultDriver;
import io.femo.http.events.*;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Ignore;
import org.junit.Test;
import java.net.URL;

import static org.junit.Assert.*;

/**
 * Created by felix on 1/19/16.
 */
public class HttpTest {

    private static JsonParser parser;

    @BeforeClass
    public static void setUp() throws Exception {
        parser = new JsonParser();
        Http.installDriver(new DefaultDriver());
    }

    @Test
    public void testHttpGet() throws Exception {
        HttpResponse response = Http.get(new URL("http://httpbin.org/get")).response();
        assertEquals("Status", 200, response.status().status());
        assertNotNull("Response String", response.responseString());
    }

    @Test
    public void testHttpGetWithParameters() throws Exception {
        HttpResponse response = Http.get(new URL("http://httpbin.org/get?param=2")).response();
        assertEquals("Status", 200, response.status().status());
        JsonObject content = parser.parse(response.responseString()).getAsJsonObject();
        JsonObject args = content.getAsJsonObject("args");
        JsonElement param = args.get("param");
        assertNotNull("Param", param);
        assertEquals("Param Value", "2", param.getAsString());
    }

    @Test
    public void testHttpStatusParsing() throws Exception {
        HttpResponse response = Http.get("http://httpbin.org/status/200").response();
        assertEquals("HTTP OK Statuscode", 200, response.statusCode());
        assertEquals("HTTP OK Status", "OK", response.status().statusMessage());
        response = Http.get("http://httpbin.org/status/404").response();
        assertEquals("HTTP Not Found Statuscode", 404, response.statusCode());
        assertEquals("HTTP Not Found Status", "NOT FOUND", response.status().statusMessage());
        response = Http.get("http://httpbin.org/status/500").response();
        assertEquals("HTTP Internal Server Error Statuscode", 500, response.statusCode());
        assertEquals("HTTP Internal Server Error Status", "INTERNAL SERVER ERROR", response.status().statusMessage());
    }

    @Test
    public void testHttpPost() throws Exception {
        HttpResponse response = Http.get(new URL("http://httpbin.org/post")).method("POST").response();
        assertEquals("Status", 200, response.status().status());
        assertNotNull("Response String", response.responseString());
    }

    @Test
    public void testHttpPostWithArguments() throws Exception {
        HttpResponse response = Http.post("http://httpbin.org/post").data("param", "2").response();
        assertEquals("Status", 200, response.statusCode());
        JsonObject content = parser.parse(response.responseString()).getAsJsonObject();
        JsonObject form = content.getAsJsonObject("form");
        JsonElement param = form.get("param");
        assertNotNull("Form Param", param);
        assertEquals("Form Param Value", "2", param.getAsString());
    }

    @Test
    public void testHttpPostWithData() throws Exception {
        HttpResponse response = Http.post("http://httpbin.org/post").entity("Value is 2").contentType("text/plain").response();
        assertEquals("Status", 200, response.statusCode());
        JsonObject content = parser.parse(response.responseString()).getAsJsonObject();
        JsonElement data = content.get("data");
        assertNotNull("Data", data);
        assertEquals("Data Value", "Value is 2", data.getAsString());
    }

    @Test
    public void testGetWithCookies() throws Exception {
        HttpResponse response = Http.get(new URL("http://httpbin.org/cookies")).cookie("Session", "abcd1234").response();
        JsonObject content = parser.parse(response.responseString()).getAsJsonObject();
        JsonObject cookies = content.getAsJsonObject("cookies");
        JsonElement cookie = cookies.get("Session");
        assertNotNull("Cookies", cookie);
        assertEquals("Cookie Value", "abcd1234", cookie.getAsString());
    }

    @Test
    public void testBasicAuthentication() throws Exception {
        HttpResponse response = Http.get("http://httpbin.org/basic-auth/test/test").basicAuth("test", "test").response();
        assertNotNull(response);
        assertEquals("Status", 200, response.statusCode());
    }

    @Test
    public void testEvents() throws Exception {
        Http.get("http://httpbin.org/get").event(HttpEventType.ALL, new HttpEventHandler() {
            @Override
            public void handle(HttpEvent event) {
                if(event.eventType() == HttpEventType.SENT) {
                    HttpSentEvent sentEvent = (HttpSentEvent) event;
                    System.out.println(sentEvent.request().requestLine());
                } else if (event.eventType() == HttpEventType.RECEIVED) {
                    HttpReceivedEvent receivedEvent = (HttpReceivedEvent) event;
                    System.out.println(receivedEvent.request().requestLine() + " - " + receivedEvent.response().statusLine());
                } else if (event.eventType() == HttpEventType.HANDLED) {
                    HttpHandledEvent handledEvent = (HttpHandledEvent) event;
                    System.out.println(handledEvent.request().requestLine() + " - " + handledEvent.response().statusLine() + " - " + (handledEvent.callbackExecuted() ? "HANDLED" : "NOT HANDLED"));
                } else {
                    System.out.println(event.eventType());
                }
            }
        }).execute();
    }
}

File: src/test/java/org/web25/http/HttpTest.kt
package org.web25.http

import com.google.gson.JsonParser
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import org.web25.http.auth.Authentication

/**
 * Created by felix on 1/19/16.
 */
internal class HttpTest {

    val http = Http()
    val parser = JsonParser()

    @Test
    fun testHttpGet() {
        val response = http.get("http://" + TestConstants.HTTP.HOST + "/get").response()
        assertEquals(200, response.status().status().toLong())
        assertNotNull(response.responseString())
    }

    @Test
    fun testHttpGetWithParameters() {
        val response = http.get("http://" + TestConstants.HTTP.HOST + "/get?param=2").response()
        assertEquals(200, response.status().status().toLong())
        val content = parser.parse(response.responseString()).asJsonObject
        val args = content.getAsJsonObject("args")
        val param = args.get("param")
        assertNotNull(param)
        assertEquals("2", param.asString)
    }

    @Test
    fun testHttpStatusParsing() {
        var response = http.get("http://" + TestConstants.HTTP.HOST + "/status/200").response()
        assertEquals(200, response.statusCode().toLong())
        assertEquals("OK", response.status().statusMessage())
        response = http.get("http://" + TestConstants.HTTP.HOST + "/status/404").response()
        assertEquals(404, response.statusCode().toLong())
        assertEquals("NOT FOUND", response.status().statusMessage())
        response = http.get("http://" + TestConstants.HTTP.HOST + "/status/500").response()
        assertEquals(500, response.statusCode().toLong())
        assertEquals("INTERNAL SERVER ERROR", response.status().statusMessage())
    }

    @Test
    fun testHttpPost() {
        val response = http.post("http://" + TestConstants.HTTP.HOST + "/post").response()
        assertEquals(200, response.status().status().toLong())
        assertNotNull(response.responseString())
    }

    @Test
    fun testHttpPostWithArguments() {
        val request = http.post("http://" + TestConstants.HTTP.HOST + "/post").entity(mapOf(Pair("param", "2")))
        val response = request.response()
        assertEquals(200, response.statusCode().toLong())
        val content = parser.parse(response.responseString()).asJsonObject
        val form = content.getAsJsonObject("form")
        val param = form.get("param")
        assertNotNull(param)
        assertEquals("2", param.asString)
    }

    @Test
    fun testHttpPostWithData() {
        val response = http.post("http://" + TestConstants.HTTP.HOST + "/post").entity("Value is 2").contentType("text/plain").response()
        assertEquals(200, response.statusCode().toLong())
        val content = parser.parse(response.responseString()).asJsonObject
        val data = content.get("data")
        assertNotNull(data)
        assertEquals("Value is 2", data.asString)
    }

    @Test
    fun testGetWithCookies() {
        val response = http.get("http://" + TestConstants.HTTP.HOST + "/cookies").cookie("Session", "abcd1234").response()
        val content = parser.parse(response.responseString()).asJsonObject
        val cookies = content.getAsJsonObject("cookies")
        val cookie = cookies.get("Session")
        assertNotNull(cookie)
        assertEquals("abcd1234", cookie.asString)
    }

    @Test
    fun testBasicAuthentication() {
        val http = Http()
        http.context.addAuthentication(Authentication.basic({"test"}, {"test"}))
        val response = http.get("http://" + TestConstants.HTTP.HOST + "/basic-auth/test/test").response()
        assertNotNull(response)
        assertEquals(200, response.statusCode().toLong())
    }

    @Test
    @Disabled("httpbin does not support digest authentication with the implemented approach")
    fun testDigestAuthentication() {
        val http = Http()
        http.context.addAuthentication(Authentication.digest({"test"}, {"test"}))
        val response = http.get("http://" + TestConstants.HTTP.HOST + "/digest-auth/auth/test/test").response()
        assertNotNull(response)
        assertEquals(200, response.statusCode().toLong())
    }

    @Test
    fun testQueryParsing(){
        val http = Http()
        val request = http.get("http://${TestConstants.HTTP.HOST}/get")
        request.query["param"] = 2
        val response = request.response()
        assertEquals(200, response.status().status().toLong())
        val content = parser.parse(response.responseString()).asJsonObject
        val args = content.getAsJsonObject("args")
        val param = args.get("param")
        assertNotNull(param)
        assertEquals("2", param.asString)
    }

    @Test
    fun testDynamicPathVariables(){
        val http = Http()
        var request = http.get("http://${TestConstants.HTTP.HOST}/status/{code}")
        request.path["code"] = 200
        var response = request.response()
        assertEquals(200, response.statusCode())

        request = http.get("http://${TestConstants.HTTP.HOST}/status/{code}")
        request.path["code"] = 404
        response = request.response()
        assertEquals(404, response.statusCode())
    }

    @Test()
    fun testNPEOnDynamicPathVars(){
        val http = Http()
        val request = http.get("http://${TestConstants.HTTP.HOST}/status/{code}")
        assertThrows<NullPointerException>(NullPointerException::class.java, {request.response()})
    }

    /*@Test
    @Throws(Exception::class)
    fun testEvents() {
        http.get("http://" + TestConstants.HTTP.HOST + "/get").event(HttpEventType.ALL, object : HttpEventHandler {
            override fun handle(event: HttpEvent) {
                if (event.eventType() === HttpEventType.SENT) {
                    val sentEvent = event as HttpSentEvent
                    println(sentEvent.request().requestLine())
                } else if (event.eventType() === HttpEventType.RECEIVED) {
                    val receivedEvent = event as HttpReceivedEvent
                    println(receivedEvent.request().requestLine() + " - " + receivedEvent.response().statusLine())
                } else if (event.eventType() === HttpEventType.HANDLED) {
                    val handledEvent = event as HttpHandledEvent
                    println(handledEvent.request().requestLine() + " - " + handledEvent.response().statusLine() + " - " + if (handledEvent.callbackExecuted()) "HANDLED" else "NOT HANDLED")
                } else {
                    println(event.eventType())
                }
            }
        }).execute()
    }*/
}